### DISCUSSION TITLE
SHA-3 512

### DISCUSSION THREAD

**Discussion by rdiwan2023** [openssl/openssl#25165] [Q&A] [answered]
Hi,
I found a SHA-3 example from demos.
https://github.com/openssl/openssl/blob/master/demos/digest/EVP_MD_demo.c

What I am seeing is when I provide the same demos example input to below link, output is not matching.
https://asecuritysite.com/encryption/sha3#:~:text=Examples%201%20Try%20%22The%20quick%20brown%20fox%20jumps,3%20Try%20%22abc%22.%20...%204%20Try%20%22abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq%22.%20

Can anyone please tell am I referring to correct demos example? if not please point a correct example of SHA3-512.

Thanks
Rohith

**Comment by kulkarniamit** [comment, marked as answer]
Padding change could be the reason for the difference:
* "_The Keccak team who created the algorithm confirmed that the actual algorithm was not changed in SHA-3. NIST only changed the padding thus it's the reason that changes the hash from Keccak._" - [Keccak vs SHA-3](https://www.cybertest.com/blog/keccak-vs-sha3)
* "_It looks like the padding change is the only difference_" - [Add support for keccak256](https://github.com/openssl/openssl/issues/13033#issuecomment-701178440)

You can see the output matching with demo example in other online tools: 
* https://emn178.github.io/online-tools/sha3_512.html
* https://codebeautify.org/sha3-512-hash-generator
* https://codifyformatter.org/sha3-512-hash-generator-online
* https://online-tool.github.io/sha3_512.html
* https://itoolkit.co/app/sha3-512-hash

> Can anyone please tell am I referring to correct demos example? if not please point a correct example of SHA3-512.

Yes, you're referring to the correct demo example for SHA3-512. The link you provided could be keccak-512 implementation.
For example the output from your link for "abc" (https://asecuritysite.com/encryption/sha3?word=abc) matches the keccak-512 implementation from OpenSSL.
```
$ echo -n "abc" | ~/bin/openssl-3.2.1/bin/openssl dgst -keccak-512
KECCAK-512(stdin)= 18587dc2ea106b9a1563e32b3312421ca164c7f1f07bc922a9c83d77cea3a1e5d0c69910739025372dc14ac9642629379540c17e2a65b19d77aa511a9d00bb96
```

**Reply by rdiwan2023** [reply]
Thanks @kulkarniamit for the details!!
